File: src/geek time/c#/week03/77. Combinations.cs
using System;
using System.Collections.Generic;

namespace week03 {
    public partial class Solution {
        private IList<IList<int>> list_77 = new List<IList<int>>();
        public IList<IList<int>> Combine(int n, int k) {
            if (n == 0 || n < k) return list_77;
            BackTrackingCombine(n, k, 1, new Stack<int>());
            return list_77;
        }
        private void BackTrackingCombine(int n, int k, int start, Stack<int> stack) {
            if (stack.Count == k) {
                // 注意这个地方一定要new一个新的，因为如果直接传stack是把引用传过去了，最终引用的结果会变成全空
                list_77.Add(new List<int>(stack));
                return;
            }
            for (var i = start; i <= n; i++) {
                stack.Push(i);
                BackTrackingCombine(n, k, i + 1, stack);
                stack.Pop();
            }
        }
    }
}

File: src/geek time/typescript/week03/77. Combinations.ts
export function combine(n: number, k: number): number[][] {
    if (!n || !k || n < k) return [];
    var ans: number[][] = [];
    recursiveCombine(n, k, ans, [], 1);
    return ans;
}

function recursiveCombine(n: number, k: number, ans: number[][], list: number[], start: number) {
    if (list.length === k) {
        ans.push(list.slice(0));
        return;
    }
    for (let i = start; i <= n; i++) {
        list.push(i);
        recursiveCombine(n, k, ans, list, i + 1);
        list.pop();
    }
}
